app/app-controllers/welcome-screen/welcomeModalController.js

```diff
@@ -90,14 +92,28 @@ function(
         var self = this;
         params = Astro.Utils.extend({forced: false}, params);
 
-        self._secureStore.get('onboarded').then(function(onboarded) {
-            if (onboarded !== 'YES' || params.forced) {
-                self.isShowing = true;
-                self.modalView.show({animated: true});
-                AppEvents.trigger(AppEvents.names.welcomeShown);
-                self._secureStore.set('onboarded', 'YES');
+        return Promise.join(
+            Application.getAppInformation(),
+            self._secureStore.get('installationID'),
+            function(appInfo, previousInstallationID) {
+                // Welcome modal should be triggered when installationID changes
+                // NOTE: appInfo.installationID appears to be different on each app run
+                // on the iOS Simulator, but on real devices they will persist.
+                if (appInfo.installationID !== previousInstallationID || params.forced) {
+                    return new Promise(function(resolve, reject) {
+                        self.isShowing = true;
+                        self.modalView.show({animated: true});
+
+                        // Promise will be resolved when welcome modal is dismissed
+                        AppEvents.on(AppEvents.names.welcomeHidden, function() {
+                            self._secureStore.set('installationID', appInfo.installationID);
+                            resolve();
```

Comment:
We `resolve();` here, but if we don't end up showing the welcome modal it looks like we never `resolve();`. That seems like the controller's `show` will never resolve then.